### ISSUE TITLE
borrow_as_ptr does not detect implicit casts due to coercions

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-clippy#14406]
### Summary

`borrow_as_ptr` should detect `&expr` in places where that is coerced to a raw pointer, and suggest `&raw const expr` instead (and similar for the mutable case)

### Lint Name

borrow_as_ptr

### Reproducer

I tried this code:

```rust
#![deny(clippy::borrow_as_ptr)]

fn main() {
    let val = 1;
    let p: *const i32 = &val;
}
```

I expected to see this happen:
the lint should fire

Instead, this happened:
it does not fire

### Version

```text
Clippy 0.1.85 (2025-02-17 4d91de4e48) on the playground
```
